### ISSUE TITLE
Uncaught TypeError: Cannot read properties of undefined

### ISSUE THREAD

**Issue by NotKoalafied** [houtianze/libby-audiobook-exporter-browser-extension#12]
I was attempting to download The Way of Kings by Brandon Sanderson, but I was unable to get the extension dialog box to show the files and download button. I checked the extensions page and discovered that there was an uncaught error, so I grabbed a screenshot from the error logs for you and placed it below. I also tried another book and had the same issue.

I am using a git-clone of the tool's master branch side loaded into Chrome version 123.0.6312.123.

![image](https://github.com/houtianze/libby-audiobook-exporter-browser-extension/assets/71797215/31049894-77d3-4940-a46e-2a881d2a6404)

**Comment by ahxxm** [comment]
cross-posting my findings for this problem https://github.com/bookbonobo/libby-download-extension/issues/161#issuecomment-2053618985

there's no `openbook.spine` now, [code pointer](https://github.com/houtianze/libby-audiobook-exporter-browser-extension/blob/master/scripts/service-worker.js#L52) here in this repo

**Comment by NotKoalafied** [comment]
I was able to verify what you're talking about. Unfortunately, I don't have the knowledge to update the code to work with the new way that Libby works. So I guess this will require Houtianze or another web dev to update the base functionality of the extension.

**Comment by mkgtu** [comment]
I acknowledge some people are having issues and I've replied to some in Reddit. But personally I've had no issues at all (so far). Just to be sure I tested today (4-13-24) both the newer and the legacy versions on my Windows PC in the Chrome, Edge, and Brave browsers. All 3 downloaded books without issues. I then tried both versions on my android (Samsung) phone in the Kiwi, Lemur, and Yandex browsers. Again, no problems at all. FYI I almost always use the Lemur android browser to download audiobooks and have never had any problems with it. I use Lemur only for Libby and nothing else and have very few other extensions or customization installed. 
  I can't deny that some have had issues recently, but I have not. If location is an issue, I'm in the San Francisco Bay Area, California, USA, using Xfinity/Comcast WiFi. 
  The 2.4 extension  installed in all these 6 browsers is the original installed from the Webstore. Though I did also sideload the GitHub unpacked version in one of them - just to see how it worked. It worked fine.

**Comment by mkgtu** [comment]
For what it's worth, there
Is another method for downloading Libby audiobooks. It's a Telegram bot (Libby Audiobook Bot). I only mention this here  because I just tried it and successfully downloaded a book. So it still works. So if Libby has changed anything it's not affecting this bot:
https://github.com/joshuatly/libby_ab_bot

**Comment by NotKoalafied** [comment]
Update:
I have now been able to remediate the issue in both Chrome and Firefox. I used different methods, so I will put some more info for each below. It's worth noting that for both methods I had to add my library card back by searching for the library and adding the number manually because copying from another device caused the issue to continue. Because I tried several things at once, I'm not sure which of them actually fixed the issue.

Firefox:
I don't use this browser regularly on the computer where I normally download, so I fully removed Firefox and tried again with a fresh installation with only the add-on for downloading being used.

Chrome:
I didn't want to fully refresh the browser, so I instead allowed 3rd party cookies and cleared the cookies for Libby.

**Comment by mkgtu** [comment]
> Update: I have now been able to remediate the issue in both Chrome and Firefox. I used different methods, so I will put some more info for each below. It's worth noting that for both methods I had to add my library card back by searching for the library and adding the number manually because copying from another device caused the issue to continue. Because I tried several things at once, I'm not sure which of them actually fixed the issue.
> 
> Firefox: I don't use this browser regularly on the computer where I normally download, so I fully removed Firefox and tried again with a fresh installation with only the add-on for downloading being used.
> 
> Chrome: I didn't want to fully refresh the browser, so I instead allowed 3rd party cookies and cleared the cookies for Libby.

So, I'm not clear. Did these things actually help? I presume they did or else why even mention them. But it wasn't clear. But the specific steps don't make a lot of sense. Are you saying the Firefox only works on a virgin installation? Maybe. I got a Reddit  reply from someone who said he got the Firefox extension to work on an old unused laptop with a new installation of Firefox with no other add-ons, but it still wouldn't work on his regular computer. And are you implying that we have to allow third party cookies to make this Chrome extension work? Is Libby now inviting in third parties to make a few extra bucks to cover expenses? In any case, my Chrome desktop blocks 3rd party cookies and this extension works fine. 
  So most of this makes no sense to me. 
   The only thing I've seen that makes sense is that Libby is rolling out some changes gradually by region and that the changes haven't hit my region yet, because for me everything still  works perfectly.

**Comment by houtianze** [comment]
Thanks for replying @mkgtu 

I tooke a quick glance of the conversation and if libby is going to make some changes at their server side, it may break things I guess.

**Comment by NotKoalafied** [comment]
@mkgtu Sorry if I was unclear. I'll clarify. My first statement says that I fixed the issue, but the reason I gave all the information about what I did is that I wasn't sure at the time which of the things I did actually fixed things and which of them did nothing. 

That said, I have now had more time to investigate and compare my results with other people. Here are the steps I have found to reliably fix the issue on my device:
1. Clear all cookies and site data from Libby.
2. When adding your card(s) back into Libby, do it manually. Copying them from another device seems to be part of what was causing the issue.
3. Attempt to use the extension again as normal.

I was also able to recreate the issue by doing the same steps except that in step 2 I would copy my card from my phone instead of adding it manually. I think that we could probably clear this up for most people by just adding a short troubleshooting section to the README.

**Comment by TrellisLore** [comment]
Clearing cookies did it for me — thank you!

**Comment by mkgtu** [comment]
> Clearing cookies did it for me — thank you!

Did it? In which browser? Firefox or chromium? Which extension, Firefox's Libby Download? Chrome's Libby Audiobook Exporter? If so, which one, 2.4 or the legacy version 1.2.2?

And what did you do, AFTER you cleared the cookies? How did you log back in?

I ask because this GitHub page is for a Chrome extension, but most of the comments in this thread are about a Firefox extension issue. It gets very confusing.

**Comment by TrellisLore** [comment]
Chrome Libby Audio exporter, version 2.4.

I logged in by inputting the library card #

**Comment by AIndoria** [comment]
Clearing cookies does nothing for me FWIW, Chrome.

**Comment by NotKoalafied** [comment]
Are you combining it with adding your card(s) manually? You should also be able to go into incognito mode to test instead of you don't want to keep clearing cookies over and over. Just make sure to enable the extension in incognito.

**Comment by AIndoria** [comment]
> Are you combining it with adding your card(s) manually? You should also be able to go into incognito mode to test instead of you don't want to keep clearing cookies over and over. Just make sure to enable the extension in incognito.

Yes, it only downloads a paltry 2 mb file (first chapter/intro), which I assume is the first mp3 file it encounters.

**Comment by NotKoalafied** [comment]
If you're getting a download/export prompt at all then this probably isn't your particular issue. Are you getting the error mentioned in the title?

**Comment by AIndoria** [comment]
I was getting the same error, and then I tried the method of clearing the
cookies. Which led to just one small file being downloaded.
